### PR TITLE
[Draft] Rebuild geojson Typescript definitions on @types/geojson

### LINKED ISSUE #1297: Turf typings don't play nicely with mapbox-gl typings
I've been working with mapbox-gl in Angular 5, and it requires using types/geojson (I am currently on v0.44.0 for the mapbox-gl typings). When I try to use your latest versions of turf, it looks like you have your own internal geojson library that you are using (it looks like since v5.1.0). So when I try to use turf helper functions such as lineString() and polygon(), it doesn't work well with mapbox-gl's typings as it sees the geojson objects as different types. Any suggestions?

### LINKED ISSUE #1658: turf-helpers: use geojson.d.ts from DefinitelyTyped
Hello :)

Why does turf-helpers ship an own version of geojson type definitions for typescript?
We could use the geojson npm package and types from https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/geojson to get better interoperatibly with other libs (for example leaflet.js).
geojson.d.ts from turf-helpers and index.d.ts from geojson are very similar and it's unnecessary to maintain them both.

### PR DESCRIPTION
**Pull request by twelch**
This PR refactors the geojson types to build on @types/geojson.

After some study, it's clear that for the most part Turf's internal geojson types are a copy of the `geojson` package, with some additions.  Where there are differences, we need either adapt the Turf code to use the new types or potentially contribute a change to @types/geojson.

Strategy:
- [x] Leave `packages/turf-helpers/lib/geojson.ts` in place as the exporter of base types for use in Turf.
- [x] Switch it to re-export base types from `geojson` package.
- [x] Keep any type additions that Turf has come to use over time (that aren't found in geojson package).  These will get dropped in a follow-on PR.
- [x] See what code breaks with the type changes and fix them. Keep blast radius small.
  - [x] features/feature collections with null geometry must now be supported.  Strategy - if function can intelligently handle it, then do so, otherwise throw with a good error message, because the code would have thrown before anyway, just with no clear error message.
- [x] Pin turf to the latest version of @types/geojson instead of '*'

Resolves #1658

Checklist:
- [ ] add docs for how *Each functions handle null geometries (e.g. coordEach, geomEach).  There is useful handling already in place!
- [ ] Add new turf sub-dependencies to package.json for packages that pick up new ones
- [ ] Add tests with null geometry to changed packages
- [ ] verify jsdocs are updated, particularly throws

Follow-on tasks to research/ticket (as observed):
- some function types don't carry on the support for null, even though the base feature/featurecollection generic do, surprised it's not complaining -- `export function featureEach<G extends GeometryObject, P = Properties>(`
- [geojson-rbush](https://github.com/DenisCarriere/geojson-rbush) package is used by multiple turf packages, which in turn imports turf types. geojson-rbush should decouple and migrate to @types/geojson.
- feature() should accept a null geometry parameter so that people can create them
- make function parameter names more accurate, at times they are overly generic or specific to what it currently accepts.  polygonDissolve(geojson), getCoord(coord).
- resolve remaining use of `any` in Typescript function signatures

----
Observations on handling features with null geometry. Conclusion - it's not exactly consistent but there is some support.  There is little documentation of null behavior.
- some Turf functions will return null if a null geometry is encountered and it can't continue or return a useful result
- some utility functions like `lineToPolygon`, `getCoord` will throw an error if null geometry encountered.  Users of this function will thus also throw - `centerOfMass`
- `coordEach` will skip null geometries
- `geomEach` will make the callback and let the user handle null geometry, and whether to keep looping

### CODE REVIEW

File: packages/turf-helpers/index.ts [hunk 1]
```diff
@@ -66,8 +66,8 @@ export type Corners = "sw" | "se" | "nw" | "ne" | "center" | "centroid";
 
 export type Lines = LineString | MultiLineString | Polygon | MultiPolygon;
 export type AllGeoJSON =
-  | Feature
-  | FeatureCollection
+  | Feature<Geometries | GeometryCollection>
+  | FeatureCollection<Geometries | GeometryCollection>
```
Comment (twelch) [inline packages/turf-helpers/index.ts:70]: This addition is to overcome the `geojson` package not providing a default interface for the generic, the user is left to specify it.  Turf had previously specified one.

geojsonDT:
```
export interface Feature<G extends GeometryObject, P = GeoJsonProperties> 
```

turf (previously):
```
export interface Feature<G = Geometry | GeometryCollection, P = Properties>
```

File: packages/turf-helpers/index.ts [hunk 2]
```diff
@@ -174,7 +174,7 @@ export let areaFactors: any = {
  *
  * //=feature
  */
-export function feature<G = Geometry, P = Properties>(
+export function feature<G extends GeometryObject, P = Properties>(
```
Comment (twelch) [inline packages/turf-helpers/index.ts:177]: This resolves the error `Type 'G' does not satisfy the constraint 'GeometryObject'.ts(2344)`

Turf implementation of Feature didn't specify that the Geometry generic had to have any certain properties (which seems in error), but `geojson` package does, saying that the Geometry generic must be like a `GeometryObject`.  The fix is for downstream generic functions feeding into it to also specify that they extend GeometryObject.

File: packages/turf-helpers/lib/geojson.ts [hunk 3]
```diff
@@ -88,158 +105,158 @@ export type Geometries =
...
+//  */
 export interface GeoJSONObject {
   // Don't include foreign members directly into this type def.
   // in order to preserve type safety.
   // [key: string]: any;
   /**
    * Specifies the type of GeoJSON object.
    */
-  type: string;
+  type: GeoJsonTypes;
```
Comment (twelch) [inline packages/turf-helpers/lib/geojson.ts:122]: makes the type more specific for compatibility

```
export type GeoJsonGeometryTypes = "Point" | "LineString" | "MultiPoint" | "Polygon" | "MultiLineString" |
    "MultiPolygon" | "GeometryCollection";
export type GeoJsonTypes = "FeatureCollection" | "Feature" | GeoJsonGeometryTypes;
```

File: packages/turf-helpers/index.ts (hunk 1 above)
Comment (twelch) [reply packages/turf-helpers/index.ts:70]: I think this default interface setting is a candidate to contribute upstream to `@types/geojson`

**Comment by JamesLMilner** [comment]
At a high level this looks like the right approach to me - I feel like where possible we should be using `@types/geojson` and if they're wrong then contributing upstream where possible.

File: packages/turf-invariant/index.ts [hunk 4]
```diff
@@ -235,7 +235,7 @@ export function collectionOf(
  */
 export function getGeom<G extends Geometries | GeometryCollection>(
   geojson: Feature<G> | G
-): G {
+): G | null {
```
Comment (twelch) [inline packages/turf-invariant/index.ts:238]: Here's the start of needing to be able to support that a Geometry can be null, which is allowed under the geojson spec, but Turf didn't support previously.

```
export interface Feature<G extends GeometryObject, P = GeoJsonProperties> extends GeoJsonObject {
    type: "Feature";
    /**
     * The feature's geometry
     */
    geometry: G | null;
```

https://datatracker.ietf.org/doc/html/rfc7946#section-3.2
```
A Feature object has a member with the name "geometry".  The value
      of the geometry member SHALL be either a Geometry object as
      defined above or, in the case that the Feature is unlocated, a
      JSON null value.
```

File: packages/turf-bezier-spline/index.ts [hunk 5]
```diff
@@ -44,16 +44,17 @@ function bezier<P = Properties>(
   const resolution = options.resolution || 10000;
   const sharpness = options.sharpness || 0.85;
 
-  const coords: [number, number][] = [];
-  const points = getGeom(line).coordinates.map((pt) => {
-    return { x: pt[0], y: pt[1] };
+  const points: { x: number; y: number }[] = [];
+  coordEach(line, (c) => {
```
Comment (twelch) [inline packages/turf-bezier-spline/index.ts:48]: coordEach will skip any null geometries

File: packages/turf-tin/index.ts [hunk 6]
```diff
@@ -52,22 +53,23 @@ export default function tin(
...
+      y: geom.coordinates[1],
+    };
+    if (z) {
+      point.z = properties[z];
+    } else if (geom.coordinates.length === 3) {
+      isPointZ = true;
+      point.z = geom.coordinates[2];
+    }
+    tinPoints.push(point);
+  });
```
Comment (twelch) [inline packages/turf-tin/index.ts:70]: Might be able to use geomReduce instead, just might not be as performant

File: packages/turf-meta/index.d.ts [hunk 7]
```diff
@@ -203,7 +206,7 @@ export function segmentReduce<Reducer extends any, P = Properties>(
     | Feature<GeometryCollection, P>
     | GeometryCollection,
   callback: (
-    previousValue?: Reducer,
+    previousValue: Reducer,
```
Comment (twelch) [inline packages/turf-meta/index.d.ts:209]: According to docs, this will always have a value, either it will be the initial value provided by the caller, or it will be the first element of the array

**Comment by mfedderly** [comment]
Is it possible to just remove packages/turf-helpers/lib/geojson.ts entirely, remove the geojson-y exports from helpers entirely and just depend on an external version of @types/geojson? That gets us out of the business of having to match every consumer's exact geojson version? This is what the mapboxgljs types do https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mapbox-gl/index.d.ts#L14

**Comment by twelch** [comment]
> Is it possible to just remove packages/turf-helpers/lib/geojson.ts entirely, remove the geojson-y exports from helpers entirely and just depend on an external version of @types/geojson? That gets us out of the business of having to match every consumer's exact geojson version? This is what the mapboxgljs types do https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mapbox-gl/index.d.ts#L14

That sounds like a reasonable goal.  That would mean each turf package that needs the types just importing fom the `geojson` package directly.  I didn't choose to bite that off initially.

I like the triple slash directive that the mapbox-gl types use to load the `geojson` types, it lets the user control which version they want.  mapbox-gl seems unique in that the types are just for external TS users, mapbox-gl internally doesn't use them.  In our case Turf uses the geojson types internally as well and needs to pin to a version right?  So while what you propose would prevent external users from importing geojson types from Turf and using them, it wouldn't avoid potential geojson version incompatibility.

**Comment by mfedderly** [comment]
> In our case Turf uses the geojson types internally as well and needs to pin to a version right? So while what you propose (removing the re-export) would prevent external users from importing geojson types from Turf and using them, it wouldn't avoid potential geojson version incompatibility.

Turf would standardize on `@types/geojson@7946.0.8` internally and we'd force that in our devDependencies using monorepolint. Users can use whatever version of @types/geojson they want, but at this point it should be compatible with the version we chose, because 7946.0.0 was released on 2017-11-28.

**Comment by twelch** [comment]
I propose to leave this first PR focused on migrating internally back to use `geojson` types.  That'll keep it small-ish and the changes across all packages will be very similar.  The dropping of lib/geojson re-exports and rewiring the imports in all the packages can be a follow-on.

File: packages/turf-line-to-polygon/index.ts [hunk 8]
```diff
@@ -81,18 +82,18 @@ function lineStringToPolygon<G extends LineString | MultiLineString>(
...
-) {
+): Feature<Polygon> {
   properties = properties
     ? properties
     : line.type === "Feature"
     ? line.properties
     : {};
   var geom = getGeom(line);
+  if (!geom || !geom.coordinates || !geom.coordinates.length)
+    throw new Error("line must contain geometry with coordinates");
```
Comment (twelch) [inline packages/turf-line-to-polygon/index.ts:93]: Here's an example of no geom throwing instead of returning null.  I kept this behavior just for consistency, but it seems like we'd want to be consistent.

File: packages/turf-boolean-contains/index.ts [hunk 9]
```diff
@@ -219,10 +219,14 @@ export function isPolyInPoly(
     return false;
   }
 
-  const coords = getGeom(feature2).coordinates;
+  const geom1 = getGeom(feature1);
+  const geom2 = getGeom(feature2);
+  if (!geom1 || !geom2) return false;
```
Comment (twelch) [inline packages/turf-boolean-contains/index.ts:224]: Perhaps all these nulls checks should be more precise `if (geom1 === null || geom2 === null)`.

File: packages/turf-line-to-polygon/index.ts (hunk 8 above)
Comment (twelch) [reply packages/turf-line-to-polygon/index.ts:93]: using throw strategy for now over returning null

File: packages/turf-helpers/index.ts (hunk 1 above)
Comment (twelch) [reply packages/turf-helpers/index.ts:70]: This is out of date, reworking

File: packages/turf-helpers/index.ts (hunk 2 above)
Comment (twelch) [reply packages/turf-helpers/index.ts:177]: actually the latest geojson types do specify a default value, resolvign
```
export interface Feature<G extends Geometry | null = Geometry, P = GeoJsonProperties> extends GeoJsonObject {
```

File: packages/turf-helpers/index.ts (hunk 1 above)
Comment (twelch) [reply packages/turf-helpers/index.ts]: The latest `geojson` types already provide a default, turf was somehow using an older version, resolving.

File: yarn.lock [hunk 10]
```diff
@@ -1231,6 +1231,11 @@
   version "7946.0.1"
```
Comment (mfedderly) [inline yarn.lock:1231]: Strange that it left the old geojson types, can you delete this block and make line 1234:

```
"@types/geojson@*", "@types/geojson@7946.0.8":
```

File: packages/turf-invariant/index.ts (hunk 4 above)
Comment (twelch) [reply packages/turf-invariant/index.ts:242]: I think I built this PR on a slightly older version of the `geojson` type package.  Turf already had a dependency on it in the top-level package.json but specified no version using only `"@types/geojson": "*"` and from what I can tell it was using a slightly older version as a result.  Once I pinned it to the latest version `7946.0.8` it picked up some nice additions.

I think it could simplify the initial migration, and we aren't forced to handle null right away.  I'm going to try and create a new smaller PR.

Specifically here is the latest `geojson` Feature type:
```
export interface Feature<G extends Geometry | null = Geometry, P = GeoJsonProperties> extends GeoJsonObject {
    type: "Feature";
    geometry: G;
    id?: string | number | undefined;
    properties: P;
}
```

Notice `null` is now added with extends instead of defined under the geometry property.  In practice this seems to allow downstream generic functions that use this to override and extend with something simpler that doesn't have null.

For example, the helper function `feature()`:
```
export function feature<G extends GeometryObject = Geometry, P = Properties>(
  geom: G,
  properties?: P,
  options: { bbox?: BBox; id?: Id } = {}
): Feature<G, P> {
```
It returns Feature, but it appears we can continue giving the generic G a lesser constraint that doesn't include `null`, so the incoming geometry just has to be like a `GeometryObject` which is to be like one of the 7 supported types `GeometryObject = Point | MultiPoint | LineString | MultiLineString | Polygon | MultiPolygon | GeometryCollection`
```.

**Comment by twelch** [comment]
Closing in favor of #2158.
